### ISSUE TITLE
Enhancement :: Support for Signatures (sign on mobile, upload image on web)

### ISSUE THREAD

**Issue by SteadyCadence** [Cadasta/cadasta-platform#1190]
Updating the issue with the new agreed format: 

### Problem Statement

Partners are planning to collect signatures from householders or witnesses for legal reasons or to verify that the data is correct. This signatures are just managed as images. 

### Context / Use Case

A typical example of this is a data collection process targeting to apply to a titling of the household. A signature of the householder is sometimes required as a legal testimony of the provided data. 

### User Stories

As a data collector, I want to collect signatures attached to specific entities (party/relationship/location) and manage them as any other image resource. 

### Description

For supporting this: 
- signatures are resources that can be associated with any of the entity types (locations, parties, relationships). Type is `signature` as specified in http://xlsform.org/. 
- To do so, we use the same syntax as we've been using for other resources in the field name: **party_resource_\***, **location_resource_\***, **tenure_resource_\***, for attaching to parties, locations and relationships as we do with the other resources, in this way the implementation effort was going to be less. It is the project manager when designing the form who will decide if the signature is associated. 

As a side note, I'd suggest to advise partners to link signatures to parties, so maybe we can include that field in the standard questionnaires (and later on in the project designer). But if in some specific cases (maybe one landowner with different parcels for instance) the partner prefers to link the signature to a location they will be able to do so. At the end this is a partners' call. 

Similarly we can include the information regarding the data collector (data collector's signature) in the standard questionnaires. 

---

(Editing to copy the requirements at the beginning of the issue, for more clarity): 

Requirement details: 
- signatures are resources that can be associated with any of the entity types (locations, parties, relationships). Type is _signature_ as specified in http://xlsform.org/. 
- To do so, we use the same syntax as we've been using for other resources in the field name: **party_resource_\***, **location_resource_\***, **tenure_resource_\***, for attaching to parties, locations and relationships as we do with the other resources, in this way the implementation effort was going to be less. It is the project manager when designing the form who will decide if the signature is associated. 
- My personal suggestion would be to advise partners to link signatures to parties, so maybe we can include that field in the standard questionnaires (and later on in the project designer). But if in some specific cases (maybe one landowner with different parcels for instance) the partner prefers to link the signature to a location they will be able to do so. 

Similarly we can include the information regarding the data collector (data collector's signature) in the standard questionnaires. 

If we create something more structured it is true that information is going to be more organized, but we are going to loose flexibility. Imagine for instance a questionnaire where not only the landowner has to sign but also the surveyor, witnesses, etc. Also there is a need for some partners to support signatures (basic functionality), maybe we could expand the feature to add more intelligence later on. 

----
XLS Forms supports signature fields through the "Appearance" column (http://xlsform.org/#appearance).  

Currently, there is no easy way to allow field collectors to grab signatures from parties when they are in the field.  

They have to use a different application or have the respondents write their signature on a piece of paper so that they can take a photo of the signature.

**Comment by wonderchook** [comment]
So I think we just need to add signature as a question type and then it should work. Since signatures are just images as far as ODK is concerned it would upload it as a resource and then we'd be good to go. 

Is there a project specifically interested in using signatures right now? Does this type of signature meet their requirements?

**Comment by SteadyCadence** [comment]
ACA Haiti has multiple signatures in their questionnaires. 

I'll try to test signatures in GeoODK with a cadasta questionnaire. 

I know that cadasta does not support the appearance column so that may cause issues. 

I'll report back.

**Comment by wonderchook** [comment]
So at the moment it doesn't work because you cant upload your questionnaire with a signature type. It should be pretty easy to fix though. Though I'm not quite sure how well our GeoODK set-up handles a ton of images, so that might be worth testing prior to adding the signature

**Comment by SteadyCadence** [comment]
I am working on another form that requires signatures.  Can we discuss this further this week?  The form I am working on does not require any photos-- only signatures.

**Comment by wonderchook** [comment]
So I am not sure discussion is really required. We can though. What is remaining for this:

1. Clear requirements written
2. Prioritized in the backlog

cc @dpalomino

**Comment by dpalomino** [comment]
@SteadyCadence @wonderchook 

Yes, it's in the backlog (id 71.00), with medium priority. Requirements are quite straight forward I think as we just need to support the _signature_ field type in the xforms which is indeed just an image. I think it shouldn't be difficult to implement. Maybe we could target as a small task for the next sprint?

**Comment by wonderchook** [comment]
@dpalomino possibly it depends on how all the other priorities line up. So that is up to you! (What do you define as a "small task"?)

**Comment by dpalomino** [comment]
Yes @wonderchook , I think it would be probably easy to implement as it is just an image that is using a different label in the xform document, so ODK interprets this as a signature and then send the image. 

Ping'ing @linzjax (thanks!) for a quick rough estimation, as she has been working on other resource stuff for xforms....

**Comment by linzjax** [comment]
So if it really is just an image, I think as long as it has the `name` column is appropriately labeled (`party_resource_<whatever>`), then it should already work as far as ODK submission goes. It would then just require updating the questionnaire model to accept that question type. Barring unforeseen complications, I don't think it would take long at all.

**Comment by dpalomino** [comment]
Great, thanks a lot @linzjax. Then I think it'd be worth to target it soon (next sprint I think?). This is useful for partners and some of them are already requesting it...

**Comment by oliverroick** [comment]
We need clear requirements for this, just supporting the field type won't do the job. 

I'm assuming partners want the signature field so people can sign that the information provided in a form is correct. Now a form usually has three entities: a location, a relationship, and a party. If we treat signatures as a resource, where does the image go? To only one of the entities? All three of them? Do partners need a way to track, which entities were collected with which form so that we can connect a signature to an entity? Do we need to store information about the data collector so that the origin of the information can be traced? 

To me, it seems that just having the image of a signature floating around as resource without any context does add much value.

**Comment by wonderchook** [comment]
In agreement with @oliverroick to be clear my comments above were related to the discussion, but did not mean we are ready to build.

**Comment by dpalomino** [comment]
Thanks @oliverroick @wonderchook for all the feedback. 

So this was my original thinking: 
- signatures are resources that can be associated with any of the entity types (locations, parties, relationships). Type is _signature_ as specified in http://xlsform.org/. 
- To do so, we use the same syntax as we've been using for other resources in the field name: **party_resource_\***, **location_resource_\***, **tenure_resource_\***, for attaching to parties, locations and relationships as we do with the other resources, in this way the implementation effort was going to be less. It is the project manager when designing the form who will decide if the signature is associated. 
- My personal suggestion would be to advise partners to link signatures to parties, so maybe we can include that field in the standard questionnaires (and later on in the project designer). But if in some specific cases (maybe one landowner with different parcels for instance) the partner prefers to link the signature to a location they will be able to do so. 

Similarly we can include the information regarding the data collector in the standard questionnaires. 

If we create something more structured it is true that information is going to be more organized, but we are going to loose flexibility. Imagine for instance a questionnaire where not only the landowner has to sign but also the surveyor, witnesses, etc. 

What do you think? 

pinging also @SteadyCadence, as she is now working with partners asking for signatures.

**Comment by wonderchook** [comment]
This makes sense to me. We should have a signature type resource though it is only an image. That way if we do reporting later we can handle them in a special way if needed.

**Comment by dpalomino** [comment]
Hi @oliverroick 

Trying to answer your questions: 

> I'm assuming partners want the signature field so people can sign that the information provided in a form is correct. Now a form usually has three entities: a location, a relationship, and a party. If we treat signatures as a resource, where does the image go? To only one of the entities? All three of them? 

Using **party_resource_\***, **location_resource_\*** or **tenure_resource_\*** the signature will be attached to the party, location or tenure. If it's data collector's signature most likely we will be attached to the location, if it's landowner's will be attached to the party, but at the end that would be a partner's call. 

> Do partners need a way to track, which entities were collected with which form so that we can connect a signature to an entity? Do we need to store information about the data collector so that the origin of the information can be traced?

I think these issues are usually addressed during the design phase of the questionnaire. So for instance, regarding data collector's information, partners usually include a couple of questions to track data collector's data (usually a drop down menu). With the naming notation mentioned before we will connect the signature with the specific entity. 

> To me, it seems that just having the image of a signature floating around as resource without any context does add much value.

Signatures will be attached to the specific entities they refer to. 

Logical next step is to include some "logic" in the project, like having a landowner's signature meaning data has been validated from landowner side, so we can have different "steps" in the property regularization. But that is beyond the management of the signatures and I think should be addressed separately in the future as a bigger project. 

In the short and medium term there are already some partners needing this solution so providing the basic implementation in the first step I think is needed. 

Hope concerns now would be a bit less, please let me know if we need more clarification here...

**Comment by dpalomino** [comment]
Adding to Sprint 18 as tentative. To be evaluated.

**Comment by dpalomino** [comment]
### Problem Statement

Partners are planning to collect signatures from householders or witnesses for legal reasons or to verify that the data is correct. This signatures are just managed as images. 

### Context / Use Case

A typical example of this is a data collection process targeting to apply to a titling of the household. A signature of the householder is sometimes required as a legal testimony of the provided data. 

### User Stories

As a data collector, I want to collect signatures attached to specific entities (party/relationship/location) and manage them as any other image resource. 

### Description

For supporting this: 
- signatures are resources that can be associated with any of the entity types (locations, parties, relationships). Type is `signature` as specified in http://xlsform.org/. 
- To do so, we use the same syntax as we've been using for other resources in the field name: **party_resource_\***, **location_resource_\***, **tenure_resource_\***, for attaching to parties, locations and relationships as we do with the other resources, in this way the implementation effort was going to be less. It is the project manager when designing the form who will decide if the signature is associated. 

As a side note, I'd suggest to advise partners to link signatures to parties, so maybe we can include that field in the standard questionnaires (and later on in the project designer). But if in some specific cases (maybe one landowner with different parcels for instance) the partner prefers to link the signature to a location they will be able to do so. At the end this is a partners' call. 

Similarly we can include the information regarding the data collector (data collector's signature) in the standard questionnaires.

**Comment by SteadyCadence** [comment]
A potential partner (big one) needs this. May need to move this up...

**Comment by jnordling** [comment]
Ok just for documentation sake......

While digging into this issues to add support for the http://xlsform.org/ signature, I discovered that 'signature' is not a data type in xlsform. What I mean is that 'text', 'image', 'select_one', or even the 'geopoint', 'geoshape' just are types. ODK represents them as a user input widget on the mobile site. Types can have 'appearance' just as they can have 'relevancy' and 'default values' all handled in the form creation side.

The 'image' type has the ability to have an 'appearance' binding tag, which in ODK will be handle by creating a widget that allows the user to sign,save and submit as a 'image'.

Long story short 'appearance' has to be added the the questioner module. 

@oliverroick @dpalomino

**Comment by dpalomino** [comment]
Thanks @jnordling! I didn't know that. 

Does it mean that when adding appearance to the questionnaire, we would be also automatically supporting the other appearance formats (like year view for dates, or multiline for text)?
http://xlsform.org/#appearance

@SteadyCadence, this is interesting...

**Comment by oliverroick** [comment]
> Does it mean that when adding appearance to the questionnaire, we would be also automatically supporting the other appearance formats (like year view for dates, or multiline for text)?

Appearance will only be supported in ODK, not in the platform; unless we make an effort to make resources available to implement more XLSForm features, which in turn is unlikely to happen in the near future.

**Comment by dpalomino** [comment]
@oliverroick 

>> Does it mean that when adding appearance to the questionnaire, we would be also automatically supporting the other appearance formats (like year view for dates, or multiline for text)?

> Appearance will only be supported in ODK, not in the platform; unless we make an effort to make resources available to implement more XLSForm features, which in turn is unlikely to happen in the near future.

Yes, that was what I meant, thanks for the clarification Oliver. Having this appearance formats in ODK (just in ODK) will be useful for some partners. Agreed that translating these appearances to the platform is not a priority (and is far less important for partners).
